rfc/system/0001-flyte-execution-tags.md

```diff
@@ -0,0 +1,124 @@
+motivation of how this feature could further improve other ecosystem projects.
+
+## 2 Motivation
+
+As a User I want to
+ - Group a certain number of executions into an experiment group - for ease of debugging and discovery.
+ - I want to mark certain executions as “blessed” or “released”. This could be done through providing a semantic version after the execution is successful 
+ - I want to group all “sub launchplan” executions with the parent execution.  
+ - External systems could group executions based on some identifiers.  
+ - Users could name their executions without having to worry about the character limits, uniqueness constraints and limited characterset.
+ - Simplify filtering of certain executions
+
+## 3 Proposed Implementation
+
+### Support for tags
+
+We propose to solve the problem of discovery by supporting arbitrary metadata association with an entity. This is similar to conceppt of “tags” as in AWS. The tags are represented as `“key”: “value”` pairs. In kuberenetes this can be represented using “labels” and “annotations”. Labels and annotations are already supported per execution as documented in - [ExecutionCreateRequest](https://docs.flyte.org/projects/flyteidl/en/latest/protos/docs/admin/admin.html#executioncreaterequest) -> [ExecutionSpec](https://docs.flyte.org/projects/flyteidl/en/latest/protos/docs/admin/admin.html#executionspec). Moreover,  every project supports default labels [Project](https://docs.flyte.org/projects/flyteidl/en/latest/protos/docs/admin/admin.html#project). Thus the final execution will have a the union of the project default labels + the user specified labels as it exists today.
+
+Currently the resultant labels are not persisted and are only applied to the
+execution in Kubernetes. As a first step, we recommend that these labels are
+persisted associated with an execution and then ListExecutions API is updated
+to return all 
+ - filtered executions by labels with supported `and` `or` queries
+ - all associated executions with every execution
+ - limit total number of labels per exection to 10-15
+
+Once this is implemented the UI and CLI can be updated to support these
+queries.
+
+### CLI Interface 
+
+A workflow or task can be executed using
+
+```bash
+pyflyte run --remote --labels k:v --labels k1:v1 test.py wf --input1=10
+```
+ (or equivalently in flytectl)
+
+flytectl and flyteremote can support filtering of executions by labels. Example
+in flytectl,
+```bash
+flytectl get execution -p flytesnacks -d development --filter.labels="k:v"
+```
+
+### UI Interface
+
+Two approaches
+
+#### Approach 1: Treat all labels the same way and allow search/filter and click based grouping
+In this approach users will get the regular executions view with all the labels
+available on each execution. The users are allowed to filter an exection simply
+by clicking on a label and then all executions are filtered by that label. 
+
+#### Approach 2: Certain label keys are treated special
```

Comment:
If this approach is chosen, I wonder whether it would be nicer for the user to do

```bash
pyflyte run --remote --group foo --experiment bar ...
```

instead of 

```bash
pyflyte run --remote --labels group:foo --labels experiment:bar ...
```

This doesn't mean that under the hood the labels mechanism couldn't be used.

Reply:
Or do we wanna create `group` and `experiment` as CLI arguments and introduce them as a concept? 🤔

Reply:
I personally prefer option 1: treat all labels the same way. Users might not want to follow the categories we deem sensible. Experiment tracking servers like Mlflow or Wandb, which also have such a tagging mechanism, simply allow users to assign arbitrary tags. I would argue that ML engineers are used to this and we should provide the same UX without imposing special naming conventions.

Only exception: execution name
I find it really helpful to have the pod names include customizable identifiers.
We have a registration script, similar to `pyflyte run` with has an `--execution_name` arg. The user provided value is appended with a random uuid, as is currently already chosen for the execution ids, and the result is checked against the execution name regex again and then passed to `FlyteRemote.execute(execution_name=...)` (already supported, see [here](https://github.com/flyteorg/flytekit/blob/26d1f290ab9e800acfd9af08595a82c3b9f962b6/flytekit/remote/remote.py#L1031)). So I wouldn't treat execution name with a pod label but the pods `metadata.name`.

Reply:
In the previous contributors synch we agreed that option 1 is preferable since we don't want to impose on users which label keys are considered special in their domain.

Reply:
@bstadlbauer @fg91 @elibixby @flixr @goyalankit Some questions for you
1. Do you prefer key-value pair tags or tags that only have key? 
2. Should we add tags to Kubernetes label?

Currently, execution spec (with labels) is serialized to byte and is stored in the execution table. it's impossible to add / delete / update tags. if we use k8s client to filtered flyteworkflow (CRD) by labels. we cannot search a execution after CR is deleted.

I have a [PR](https://github.com/flyteorg/flyteadmin/pull/571) that adds tags table. it allows us easily add / update / delete tags, and even attach tags to task / workflow / project. however, it's not key-value pair tags for now. If we decide to use key-value pair tags, I just need to add a new column to the tags table and update the query. I'd like to know your thought first. 

btw, the current implementation works with both Mysql and Postgres.

Reply:
My $0.02 is let's keep it simple and support what you call key-only tags.  

A person can 'hack' this to resemble key-value if needed (ie 'costcenter-12'), but we don't need to manage that complexity on the back end or in the UI when we get to figuring out how to let folks use tags to sort/group things.

Reply:
* In my opinion key-only tags are perfectly fine and what ML engineers are used to from experiment tracking servers
* > Should we add tags to Kubernetes label

    I think being able to add/delete/update tags after the execution has already started or ended is an important feature. User story: an experiment is training/trained really well and I want to mark it for later. This is something that is not known when starting the execution. But updating/deleting/adding tags when the execution is already running would mean that the k8s labels are not in sync with what is stored in the tags table. I'd therefore say that I wouldn't apply the tags as labels to k8s.

Reply:
Sorry for the late response here but agreed with what's been said above. Key only tags would also solve all our usecases 👍 

> I think being able to add/delete/update tags after the execution has already started or ended is an important feature 
+1 to this and the reasoning of not applying those to k8s